src/Swift.Runtime/src/Library/Arc.cs

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Swift.Runtime;
+
+/// <summary>
+/// Arc is a class containing p/invokes for Swift Automatic Reference Counting and memory management.
+/// </summary>
+public static class Arc {
+    const string swiftCore = "/usr/lib/swift/libswiftCore.dylib";
+
+    /// <summary>
+    /// Retain a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    [DllImport(swiftCore, CallingConvention = CallingConvention.Cdecl)]
+    static extern void swift_retain(IntPtr p);
+
+    /// <summary>
+    /// Retain a heap-allocated Swift object
+    /// </summary>
+    /// <param name="p">Pointer to an unmanaged Swift object, must be non-null.</param>
+    /// <returns>The pointer passed in.</returns>
+    /// <exception cref="ArgumentOutOfRangeException">Throws if p is null</exception>
+    public static IntPtr Retain(IntPtr p)
+    {
+        if (p == IntPtr.Zero)
+            throw new ArgumentOutOfRangeException(nameof(p));
```

Comment:
This might just be me but:
If any of these argument checks fails and an exception is thrown, the message that the developer will see would be like:
>System.ArgumentOutOfRangeException: 'Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'p')'.

Since there is little maintenance involved here, can the parameter name be a little more descriptive?